### PR TITLE
Implement total-spin observable in PennyLane-QChem

### PR DESCRIPTION
**Pull request by agran2018**
**Context:**
Implementation of observables, other than the electronic Hamiltonian, in PL-QChem to compute molecular properties.

**Description of the Change:**
This PR adds the new module ``obs`` containing the functions to build the total-spin (S^2) many-body operator and make it a PennyLane observable.

**Benefits:**
Provides a general framework to build operators to simulate the properties of interest of a many-body quantum systems.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: doc/conf.py [hunk 1]
```diff
@@ -129,6 +129,7 @@ def __getattr__(cls, name):
     MOCK_MODULES = [
         'pennylane_qchem',
         'pennylane_qchem.qchem',
+        'pennylane_qchem.obs',
```
Comment (josh146) [inline doc/conf.py:132]:
```suggestion
```

File: qchem/pennylane_qchem/qchem/obs.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+import os
+import numpy as np
+from openfermion.hamiltonians import MolecularData
+from openfermion.ops import FermionOperator
+from openfermion.transforms import bravyi_kitaev, jordan_wigner
+
+from . import structure
+
+
+def s2_me_table(sz, n_spin_orbs):
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:27]: I'm trying to think of a function name that better communicates to users/developers what this function does. `table` is probably redundant, we should instead emphasize the 'two particle spin' and 'matrix elements' part.

* `spin_matrix_elements`

* `spin2_matrix_elements`

* `two_particle_spin_elements`

Any other thoughts? 🤔

File: qchem/pennylane_qchem/qchem/obs.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    r"""Generates the table of the matrix elements
+    :math:`\langle \alpha, beta \vert \hat{s}_1 \cdot \hat{s}_2 \vert \gamma, \delta \rangle`
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:29]:
```suggestion
    :math:`\langle \alpha, \beta \vert \hat{s}_1 \cdot \hat{s}_2 \vert \gamma, \delta \rangle`
```

File: qchem/pennylane_qchem/qchem/obs.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    The matrix elements are evaluated using the expression,
+
+    .. math:
+
+        \langle \alpha, \beta \vert \hat{\bf{s}}_1 \cdot \hat {\bf{s}}_2
+        \vert \gamma, \delta \rangle = && \delta_{\alpha,\delta} \delta_{\beta,\gamma} \times \\
+        && \left( \frac{1}{2} \delta_{m_\alpha, m_\delta+1} \delta_{m_\beta, m_\gamma-1}
+        + \frac{1}{2} \delta_{m_\alpha, m_\delta-1} \delta_{m_\beta, m_\gamma+1}
+        + m_\alpha m_\beta \delta_{m_\alpha, m_\delta} \delta_{m_\beta, m_\gamma} \right),
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:40]: @agran2018, this is not showing up for me when I build the docs 🤔

File: qchem/pennylane_qchem/qchem/obs.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+    where :math:`\alpha` and :math:`m_\alpha` refer to the quantum numbers of the spatial
+    :math:`\varphi_\alpha({\bf r})` and spin :math:`\chi_{m_\alpha}(s_z)` wave functions,
+    respectively, of the single-particle state :math:`\vert \alpha \rangle`.
+
+    **Example**
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:46]: We have settled on a new standard for docstrings; examples now go after the Args: and Returns:. See https://pennylane.readthedocs.io/en/latest/development/guide/documentation.html#functions-and-methods

File: qchem/pennylane_qchem/qchem/obs.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+     [ 1.    1.    1.    1.    0.25]
+     [ 0.    1.    0.    1.    0.5 ]
+     [ 1.    0.    1.    0.    0.5 ]]
+
+    Args:
+        sz (array[float]): spin-projection quantum number of the spin-orbitals
+        n_spin_orbs (int): number of spin orbitals
+
+    Returns:
+        array: NumPy array with the table of matrix elements
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:62]: Might also be good to explain the convention used in the returned table. E.g., which elements are the indices, which are the values.

File: qchem/pennylane_qchem/qchem/obs.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            filename=os.path.join(hf_data.strip(), mol_name.strip())
+        )
+        n_electrons = hf_elect_struct.n_electrons
+    else:
+        n_electrons = n_active_electrons
+
+    n_spin_orbs = 2 * len(active_indices)
+    sz = np.where(np.arange(n_spin_orbs) % 2 == 0, 0.5, -0.5)
+
+    return s2_me_table(sz, n_spin_orbs), 3 / 4 * n_electrons
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:164]: Nice :)

File: qchem/pennylane_qchem/qchem/obs.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+
+def observable(me_table, init_term=0, mapping="jordan_wigner"):
+
+    r"""Builds the many-body observable whose expectation value can be
+    measured in PennyLane
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:170]:
```suggestion
    r"""Builds the many-body observable whose expectation value can be
    measured in PennyLane.
```

File: qchem/pennylane_qchem/qchem/obs.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        \vert \alpha \rangle ~ \hat{n}_\alpha +
+        \sum_{\alpha, \beta ~ \in ~ \mathrm{active~space}} \langle \alpha \vert \hat{\mathcal{A}}
+        \vert \beta \rangle ~ \hat{c}_\alpha^\dagger\hat{c}_\beta,
+
+        \hat B = \frac{1}{2} \left\{ \sum_{\alpha, \beta \leq 2N_\mathrm{docc}}
+        \langle \alpha, \beta \vert \hat{\mathcal{B}} \vert \beta, \alpha \rangle
+        ~ \hat{n}_\alpha \hat{n}_\beta + \sum_{\alpha, \beta, \gamma, \delta ~
+        \in ~ \mathrm{active~space}} \langle \alpha, \beta \vert \hat{\mathcal{B}}
+        \vert \gamma, \delta \rangle ~ \hat{c}_{\alpha}^\dagger \hat{c}_{\beta}^\dagger
+        \hat{c}_{\gamma} \hat{c}_{\delta} \right\}.
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:189]: equation alignment seems a bit off here

File: qchem/pennylane_qchem/qchem/obs.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    + (0.375) [Z<Wires = [3]>]
+    + (-0.375) [Z<Wires = [2]> Z<Wires = [3]>]
+    + (0.125) [Y<Wires = [0]> X<Wires = [1]> Y<Wires = [2]> X<Wires = [3]>]
+    + (0.125) [Y<Wires = [0]> Y<Wires = [1]> X<Wires = [2]> X<Wires = [3]>]
+    + (0.125) [Y<Wires = [0]> Y<Wires = [1]> Y<Wires = [2]> Y<Wires = [3]>]
+    + (-0.125) [Y<Wires = [0]> X<Wires = [1]> X<Wires = [2]> Y<Wires = [3]>]
+    + (-0.125) [X<Wires = [0]> Y<Wires = [1]> Y<Wires = [2]> X<Wires = [3]>]
+    + (0.125) [X<Wires = [0]> X<Wires = [1]> X<Wires = [2]> X<Wires = [3]>]
+    + (0.125) [X<Wires = [0]> X<Wires = [1]> Y<Wires = [2]> Y<Wires = [3]>]
+    + (0.125) [X<Wires = [0]> Y<Wires = [1]> X<Wires = [2]> Y<Wires = [3]>]
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:229]: Nice example! Quick comment though:

* @mariaschuld has temporarily removed the Wires class from Observables, so the above example should be changed to `(0.125) [X0 Y1 X2 Y3]` etc.

* Even when the Wires class is re-introduced, the above is a bug --- we don't want the output to look like that! So the `Hamiltonian` class will be modified to make sure it continues to print as before.

File: qchem/pennylane_qchem/qchem/obs.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+    Args:
+        me_table (array[float]): Numpy array with the table of matrix elements.
+            For a single-particle operator this array will have shape
+            ``(me_table.shape[0], 3)`` with each row containing the indices
+            :math:`alpha`, :math:`beta` and the matrix element :math:`\alpha \vert
+            \hat{\mathcal{A}}\vert \beta \rangle`. For a two-particle operator this
+            array will have shape ``(me_table.shape[0], 5)`` with each row containing
+            the indices :math:`alpha`, :math:`beta`, :math:`gamma`, :math:`delta` and
+            the matrix elements :math:`\langle \alpha, \beta \vert \hat{\mathcal{B}}
+            \vert \gamma, \delta \rangle`.
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:240]:
```suggestion
            :math:`\alpha`, :math:`\beta` and the matrix element :math:`\langle \alpha \vert
            \hat{\mathcal{A}}\vert \beta \rangle`. For a two-particle operator this
            array will have shape ``(me_table.shape[0], 5)`` with each row containing
            the indices :math:`\alpha`, :math:`\beta`, :math:`\gamma`, :math:`\delta` and
            the matrix elements :math:`\langle \alpha, \beta \vert \hat{\mathcal{B}}
            \vert \gamma, \delta \rangle`.
```

File: qchem/pennylane_qchem/obs.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+    In the latter equations :math:`N_\mathrm{docc}` denotes the doubly-occupied orbitals,
+    if any, not included in the active space and
+    :math:`\langle \alpha \vert \hat{\mathcal{A}} \vert \beta \rangle` and
+    :math:`\langle \alpha, \beta \vert\hat{\mathcal{B}} \vert \gamma, \delta \rangle`
+    are the matrix elements of the one- and two-particle operators
+    :math:`\hat{\mathcal{A}}` and :math:`\hat{\mathcal{B}}`, respectively.
+
+    The function utilizes tools of `OpenFermion <https://github.com/quantumlib/OpenFermion>`_
+    to buil the second-quantized operator and map it to basis of Pauli matrices via the
```
Comment (soranjh) [inline qchem/pennylane_qchem/obs.py]:
```suggestion
    to build the second-quantized operator and map it to basis of Pauli matrices via the
```

File: qchem/tests/test_s2_obs.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+        ("h2_pyscf", 2, None, me, 1.5),
+        ("lih", 2, 2, me, 1.5),
+        ("lih", None, 3, me_lih, 3.0),
+    ],
+)
+def test_get_s2_matrix_elements(
+    mol_name, n_act_elect, n_act_orb, s2_me_exp, init_term_exp, tol
+):
+    r"""Test that the table of matrix elements and the term use to initialize the
+    FermionOperator are computed corrrectly for different active spaces."""
```
Comment (soranjh) [inline qchem/tests/test_s2_obs.py:161]:
```suggestion
    FermionOperator are computed correctly for different active spaces."""
```

File: qchem/pennylane_qchem/qchem/obs.py (hunk 2 above)
Comment (soranjh) [reply qchem/pennylane_qchem/qchem/obs.py:27]: Another option is just `spin2_matrix`.

File: qchem/pennylane_qchem/qchem/obs.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+    r"""Generates the table of the matrix elements
+    :math:`\langle \alpha, beta \vert \hat{s}_1 \cdot \hat{s}_2 \vert \gamma, \delta \rangle`
+    of the two-particle spin operator :math:`\hat{s}_1 \cdot \hat{s}_2`.
+
+    The matrix elements are evaluated using the expression,
+
+    .. math:
+
+        \langle \alpha, \beta \vert \hat{\bf{s}}_1 \cdot \hat {\bf{s}}_2
+        \vert \gamma, \delta \rangle = && \delta_{\alpha,\delta} \delta_{\beta,\gamma} \times \\
```
Comment (soranjh) [inline qchem/pennylane_qchem/qchem/obs.py:37]: Is the symbol :math:`\delta` used for two different purposes?

File: qchem/pennylane_qchem/qchem/obs.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+        + m_\alpha m_\beta \delta_{m_\alpha, m_\delta} \delta_{m_\beta, m_\gamma} \right),
+
+    where :math:`\alpha` and :math:`m_\alpha` refer to the quantum numbers of the spatial
+    :math:`\varphi_\alpha({\bf r})` and spin :math:`\chi_{m_\alpha}(s_z)` wave functions,
+    respectively, of the single-particle state :math:`\vert \alpha \rangle`.
+
+    **Example**
+
+    >>> sz = np.array([0.5, -0.5])
+    >>> print(s2_me_table(sz, 2))
```
Comment (soranjh) [inline qchem/pennylane_qchem/qchem/obs.py:49]: maybe:
`n_spin_orbs = 2`
`print(s2_me_table(sz, n_spin_orbs))`

File: qchem/pennylane_qchem/qchem/obs.py (hunk 2 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:27]: I think the simpler way would be to call it ``spin2_coeffs``. Matrix elements may be confusing as you can also compute matrix elements of the full operator, indeed, this is what it is mostly done.

File: qchem/pennylane_qchem/qchem/obs.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        sz (array[float]): spin-projection quantum number of the spin-orbitals
+        n_spin_orbs (int): number of spin orbitals
+
+    Returns:
+        array: NumPy array with the table of matrix elements
+    """
+
+    if sz.size != n_spin_orbs:
+        raise ValueError(
+            "Size of 'sz' must be equal to 'n_spin_orbs'; got {}".format(sz.size)
```
Comment (soranjh) [inline qchem/pennylane_qchem/qchem/obs.py:67]: Maybe:
"Size of 'sz' must be equal to 'n_spin_orbs'; got {} for 'sz' size".format(sz.size)

File: qchem/pennylane_qchem/qchem/obs.py (hunk 5 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:46]: Oh, that makes sense. Thanks @josh146

File: qchem/pennylane_qchem/qchem/obs.py (hunk 6 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:62]: I think this is explained in one of the functions below, but will be happy to add it also here.

File: qchem/pennylane_qchem/qchem/obs.py (hunk 4 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:40]: will check this

File: qchem/pennylane_qchem/qchem/obs.py (hunk 9 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:189]: yes, thanks.

File: qchem/pennylane_qchem/qchem/obs.py (hunk 10 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:229]: Nice!

File: qchem/pennylane_qchem/qchem/obs.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+
+    # we only care about indices satisfying the following boolean mask
+    mask = np.logical_and(alpha // 2 == delta // 2, beta // 2 == gamma // 2)
+
+    # diagonal elements
+    diag_mask = np.logical_and(sz[alpha] == sz[delta], sz[beta] == sz[gamma])
+    diag_indices = np.argwhere(np.logical_and(mask, diag_mask))
+    diag_values = (sz[alpha] * sz[beta]).flatten()
+
+    diag = np.vstack([diag_indices.T, diag_values]).T
```
Comment (nquesada) [inline qchem/pennylane_qchem/qchem/obs.py:85]: I have to admit that I always find very hard to read what all these reshaping does. I understand is faster but would not be able to tell you what happened in the last 15 lines.

File: qchem/tests/test_observable.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        np.array([[0.0, 0.0, 1.0, 0.5], [1.0, -0.5]]),
+        np.array([[0.0, 0.0, 1.0, 2.0, 0.5], [1.0, -0.5]]),
+        np.array([[0.0, 0.0, 1.0, 2.0, 3.0, 0.5], [1.0, 0.0, -0.5]]),
+    ],
+)
+def test_exceptions_observable(
+    me_table, message_match="expected entries of 'me_table' to be of shape"
+):
+    """Test that the 'observable' function throws an exception if the
+    array containing the matrix elements has illegal shapes."""
```
Comment (nquesada) [inline qchem/tests/test_observable.py:115]:
```suggestion
    array containing the matrix elements has incorrect shapes."""
```

File: qchem/pennylane_qchem/qchem/obs.py (hunk 14 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:37]: Soran, this is for the Kronecker delta function, if renders properly in the docs there must be no confusion about it.

File: qchem/pennylane_qchem/qchem/obs.py (hunk 10 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:229]: Done

File: qchem/pennylane_qchem/qchem/obs.py (hunk 17 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py:88]: Only for @josh146 :-)

**Comment by agran2018** [comment]
> Nice work @agran2018! A couple of comments and thoughts:

Thanks @josh146.

> 
> * The file `qchem/tmplv5rrgvd` keeps being accidentally committed. We should probably add `qchem/tmp*` to the `.gitignore` file.

I noticed that this is done, right?.

> * In order to get the new `qml.qchem.obs` module to show up in the docs, I had turn `pennylane_qchem.qchem` from a module into a _package_ that contains the submodules `pennylane_qchem.qchem.structure` (the electronic structure functions) and `pennylane_qchem.qchem.obs`.

I see. I am not sure about the name ``structure`` but I guess we can always refine this later, right?. For instance, the way I see it is that both ``structure`` and ``observable`` contains functions that are needed to perform quantum electronic structure simulations e.g., VQE. We might want to have in the future a module that implements electronic structure simulations USING the core functions in ``structure`` and available functionalities in ``obs``.

On the other hand I guess this new arrangement will conflict with the this PR https://github.com/XanaduAI/pennylane/pull/679 ?. Which the best way to proceed in this case?. Should we get this merged first?. Then open a new one to accommodate  https://github.com/XanaduAI/pennylane/pull/679 ?. Bit confused here.

>   The functions `structure` are imported into `pennylane_qchem/qchem/__init__`, so remain 'top-level'. E.g., you can continue to do `qml.qchem.generate_hamiltonian`.
> * The functions in the `obs` module are not imported into `__init__`, so must be accessed via `qml.qchem.obs.observable`. Is this intentional? At the very least, I feel that `observable` should be 'top-level`, so that you can do `qml.qchem.observable`.

Yes, I agree. These functions should be at the top level as well.

> * You will need to `black -l 100 qchem/` to make sure that the new qchem subpackage is properly blacked (I noticed a lot of trailing spaces).
Will do. I was thinking to do this in a separate PR but good to do now that there is a restructuring going on.

File: qchem/pennylane_qchem/qchem/obs.py (hunk 9 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py]: done

File: qchem/pennylane_qchem/qchem/obs.py (hunk 4 above)
Comment (agran2018) [reply qchem/pennylane_qchem/qchem/obs.py]: Working now!.

**Comment by josh146** [comment]
> I see. I am not sure about the name structure but I guess we can always refine this later, right?. 

Yep! I just needed to create separate files for them, as I couldn't get the documentation to build otherwise. Feel free to rename the file.

> On the other hand I guess this new arrangement will conflict with the this PR #679 ?. Which the best way to proceed in this case?. Should we get this merged first?. Then open a new one to accommodate #679 ?. Bit confused here.

It should be fine 🙂 Once one of the PRs is merged, there will be a merge conflict in the other (which should be relatively easy to resolve).

File: qchem/pennylane_qchem/qchem/obs.py (hunk 17 above)
Comment (josh146) [reply qchem/pennylane_qchem/qchem/obs.py:88]: No I agree, this function is very hard to read. I wonder how to improve it.

File: qchem/pennylane_qchem/qchem/obs.py [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+    off_diag_indices = np.argwhere(off_diag_mask)
+    off_diag_values = np.full([len(off_diag_indices)], 0.5)
+
+    off_diag = np.vstack([off_diag_indices.T, off_diag_values]).T
+
+    # combine the off diagonal and diagonal tables into a single table
+    return np.vstack([diag, off_diag])
+
+
+def get_spin2_matrix_elements(mol_name, hf_data, n_active_electrons=None, n_active_orbitals=None):
```
Comment (josh146) [inline qchem/pennylane_qchem/qchem/obs.py:104]: new function names are much better!

File: qchem/CHANGELOG.md [hunk 20]
```diff
@@ -2,6 +2,13 @@
 
 <h3>New features since last release</h3>
 
+* The new module ``obs`` has been added to build many-body operators
```
Comment (josh146) [inline qchem/CHANGELOG.md:5]: Will need to be moved into the v0.11 section!

**Pull request merged**